lib/puppet/util/json.rb

```diff
@@ -0,0 +1,65 @@
+module Puppet::Util
+  module Json
+    class ParseError < StandardError
+    end
+
+    begin
+      require 'multi_json'
+      # Force backend detection before attempting to use the library
+      # or load any other JSON libraries
+      MultiJson.default_adapter
+
+      # Preserve core type monkey-patching done by the built-in JSON gem
+      require 'json'
+    rescue LoadError
+      require 'json'
+    end
+
+    # These methods do similar processing to the fallback implemented by MultiJson
+    # when using the built-in JSON backend, to ensure consistent behavior
+    # whether or not Puppet::Util::Json can be loaded.
+    def self.load(string, options = {})
+      if defined? MultiJson
+        begin
+          MultiJson.load(string, options)
+        rescue MultiJson::ParseError => e
+          raise Puppet::Util::Json::ParseError, e.cause
```

Comment:
Do we want to do more here about preserving the original cause? This saves the message and ruby will automatically save the cause chain but the backtrace will end here. MultiJson does something more involved to make sure the immediately returned exception has the full trace, not sure if we care.

Reply:
It doesn't look too difficult to do something more like what they are doing. I'm happy to add it.

Reply:
Updated.